refactor(docs-infra): migrate from tslint to eslint

dario-piotrowicz opened the pull request:
migration to eslint as tslint has been deprecated

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [x] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

`ng lint` runs the deprecated tslint

Issue Number: N/A

## What is the new behavior?
`ng lint` runs eslint

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

dario-piotrowicz commented:
Hi @gkalpak :slightly_smiling_face: 

So, the migration is looking.... good?... hopefully? :sweat_smile: 

![Screenshot at 2021-07-11 10-41-04](https://user-images.githubusercontent.com/61631103/125190352-91f71d00-e234-11eb-828f-3a096efbc555.png)

I opened this draft PR so start a bit of a discussion and see how you see this, I hope you don't mind :slightly_smiling_face:

dario-piotrowicz commented:
First of all, what do you think of the incompatible typescript version? is this something we can accept or it's a deal breaker? :thinking: 

(I would not want to downgrade the current typescript just to accommodate eslint)

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,100 @@
+        "no-tabs": "error",
+        "no-restricted-syntax": [
+          "error",
+          {
+              "selector": "CallExpression[callee.object.name='console'][callee.property.name=/^(debug|info|time|timeEnd|trace)$/]",
+              "message": "Unexpected property on console object was called"
+          },
```
I needed to put this here as the eslint rule `no-console` didn't seem to be working... (no idea why... maybe it could be because of the typescript version mismatch?)

dario-piotrowicz commented on aio/src/404-body.html:
```diff
@@ -27,7 +27,7 @@
             <div class="content">
               <div class="nf-container l-flex-wrap flex-center">
-                <img src="assets/images/support/angular-404.svg" width="300" height="300" />
+                <img src="assets/images/support/angular-404.svg" width="300" height="300" alt="not-found"/>
```
Fixes
![Screenshot at 2021-07-11 10-49-41](https://user-images.githubusercontent.com/61631103/125190597-e222af00-e235-11eb-87d7-b5a00c532dc1.png)

dario-piotrowicz commented on aio/src/app/app.component.spec.ts:
```diff
@@ -589,7 +589,7 @@ describe('AppComponent', () => {
 
     describe('restrainScrolling()', () => {
-      const preventedScrolling = (currentTarget: object, deltaY: number) => {
+      const preventedScrolling = (currentTarget: { scrollTop: number, scrollHeight?: number, clientHeight?: number }, deltaY: number) => {
```
here and in the other instances of `:object`, I changed those as `object` seems to have issues
![Screenshot at 2021-07-11 10-54-29](https://user-images.githubusercontent.com/61631103/125190747-98869400-e236-11eb-9650-f5189f2aa091.png)

dario-piotrowicz commented on aio/src/app/shared/search-results/search-results.component.ts:
```diff
@@ -1,3 +1,5 @@
+/* eslint-disable no-shadow */
+// not sure why the above eslint-disable comment is necessary (eslint bug?)
```
This is rather strange (again, may this be because of the typescript version mismatch? :thinking: )

eslint complains that `SearchState` gets shadowed... but it does not seem to be the case

rafaelss95 replied:
Have you tried to use [**`@typescript-eslint/no-shadow`**](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-shadow.md) instead?

dario-piotrowicz replied:
good point, I'll give it a shot thanks! :slightly_smiling_face:

dario-piotrowicz replied:
spot on! :smiley:

dario-piotrowicz commented on aio/src/app/custom-elements/resource/resource-list.component.html:
```diff
@@ -8,7 +8,7 @@
   <div class="showcase">
     <div *ngFor="let subCategory of selectedCategory?.subCategories">
-      <a id="{{subCategory.id}}"></a>
+      <span id="{{subCategory.id}}"></span>
```
fixes 
![Screenshot at 2021-07-11 11-03-44](https://user-images.githubusercontent.com/61631103/125190950-cd471b00-e237-11eb-9e42-eb63017dee43.png)

I had a look at the site and I am not sure the the `a` was necessary there

I guess the `id` is used for fragment navigation, but not sure why the element needed to be an anchor

dario-piotrowicz replied:
Also I put the rule there to `"warn"`, not sure about it... should all our rules be `"error"`?

gkalpak replied:
I'd rather have errors over warnings whenever possible.

dario-piotrowicz replied:
got it, I'll see all rules to error :slightly_smiling_face: 

___

about the `h3` you are totally right... but yeah I am bit confused... why wasn't that done in the first place? are we missing something? :thinking: 

it's been like that for a long time: https://github.com/angular/angular/commit/196203f6d76ac5240d1485dbd9701bfb026eab56

but actually if you look at the commit also the categories were implemented in the same way, but those have been updated to `a` tags altogether around a year ago: https://github.com/angular/angular/commit/e672b1f2ac42b3a3b59f6020457318f8e1aa78dd 

I guess it is fine to change it, but yeah I am a bit confused on why it's been implemented that way :thinking: 

___

Actually.... I don't think any id is necessary/used here :confused: , we may want to clean those up (after this PR maybe? :) )
For example check this out: https://angular.io/resources#data-libraries (the url gets changed in the onInit regardless and the fragment gets ignored)

gkalpak replied:
Yeah, they are probably unusable. But let's keep them in this PR, since removing them is outside the scope of it. (Moving the IDs to the `h3` elements should preserve the previous behavior (broken or not), I think it is a good way to pass linting :smiley:)

dario-piotrowicz replied:
yeah agreed! I'm on it :slightly_smiling_face:

rafaelss95 commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,100 @@
+              "selector": "CallExpression[callee.object.name='console'][callee.property.name=/^(debug|info|time|timeEnd|trace)$/]",
+              "message": "Unexpected property on console object was called"
+          },
+          {
+            "selector": "CallExpression[callee.name='fdescribe']",
+            "message": "Don't keep jasmine focus methods."
+          },
+          {
+            "selector": "CallExpression[callee.name='fit']",
+            "message": "Don't keep jasmine focus methods."
+          }
```
I know it might have come from the automatic conversion, but you can condense these two into:

```suggestion
          {
            "selector": "CallExpression[callee.name=/^(fdescribe|fit)$/]",
            "message": "Don't keep jasmine focus methods."
          }
```

dario-piotrowicz replied:
good point! :slightly_smiling_face: 

This actually was not automated but it myself :sweat_smile:  (as the tslint ban rule doesn't have a 1-to-1 alternative), I did it separately as I thought that the two rules had each a different/specific message, I did notice that the error message was the same but forgot to combine the rules! thanks for noticing it! :slightly_smiling_face:

rafaelss95 replied:
😮 Maybe you could have follow the steps described [here](https://github.com/angular-eslint/angular-eslint#migrating-an-angular-cli-project-from-codelyzer-and-tslint) 😄

dario-piotrowicz replied:
yeah I did!

sorry if my comment was unclear :stuck_out_tongue_winking_eye: , I did follow that, what I meant is that the schematics failed to the tslint ban rules one rule (which did what I am doing here), they just provided such line in the log file: 
```
* tslint-to-eslint-config does not know the ESLint equivalent for TSLint's "ban".
```

so I wrote this rule myself :slightly_smiling_face: 
(the schematics did an awesome job but after those I did need to do quite some manual adjustments and stuff)

dario-piotrowicz commented on aio/src/app/custom-elements/resource/resource-list.component.ts:
```diff
@@ -4,7 +4,7 @@ import { Category } from './resource.model';
 import { LocationService } from 'app/shared/location.service';
 
-/* tslint:disable:template-accessibility-elements-content */
+/* eslint-disable @angular-eslint/template/accessibility-elements-content */
```
ah! yeah I think you're right! thanks!!! :slightly_smiling_face:

gkalpak reviewed (changes requested):
Awesome work (as usual), @dario-piotrowicz ❤️
Just a few minor comments, but this is looking very promising :smiley:

dario-piotrowicz commented:
> Awesome work (as usual), @dario-piotrowicz 
> Just a few minor comments, but this is looking very promising

Thanks a lot for the kind words!!!! :slightly_smiling_face: 

Well, your comments don't seem to be "just a few" :sweat_smile: (I'm going to check them out now!),
I really hope I am not burdening you with this :sweat:
thank you so very much for taking the time to review the PR :smiling_face_with_three_hearts: 
___

By the way, could you comment on the typescript version warning? don't you see it as a problem here? :slightly_smiling_face:

dario-piotrowicz commented on aio/angular.json:
```diff
@@ -6,7 +6,8 @@
     },
-    "analytics": false
+    "analytics": false,
+    "defaultCollection": "@angular-eslint/schematics"
```
:open_mouth: wow yeah!

sorry the sneaky eslint schematics must have put that there when I run them for the automated conversion process, I didn't noticed this

I'll get rid of it :slightly_smiling_face:

dario-piotrowicz commented on aio/angular.json:
```diff
@@ -196,16 +197,11 @@
-              "**/node_modules/**"
+            "lintFilePatterns": [
+              "src/**/*.ts",
+              "src/**/*.html"
```
yeah good point! :slightly_smiling_face: :+1:

alan-agius4 replied:
`tests/` is missing from the files to lint.

dario-piotrowicz replied:
the addition of the tests one reveals other files that I need to amend there :sweat: 

I'll get on with it after fixing the other stuff, good catch! :slightly_smiling_face:

dario-piotrowicz commented on aio/package.json:
```diff
@@ -105,6 +105,11 @@
+    "@angular-eslint/eslint-plugin": "12.2.2",
+    "@angular-eslint/eslint-plugin-template": "12.2.2",
+    "@angular-eslint/schematics": "12.2.2",
+    "@angular-eslint/template-parser": "12.2.2",
```
:sweat: 

I am no expert... they do seem necessary right? the only one I guess we could remove if we wanted to are the schematics I'd imagine

alan-agius4 replied:
`@angular-eslint/schematics` shouldn't be needed.

alan-agius4 replied:
Tbh, to keep the same lint behaviour we might need to add additional dependencies. IMHO, this is one of the biggest drawbacks of Eslint.

dario-piotrowicz replied:
well, eslint is a more general purpose tool whilst tslint was typescript specific, so it kind of bundled for us everything we needed, things tat now we need to bring in with plugins... yeah not the best, but I think it is already quite impressive that eslint can manage tyepscript at all in the first place! :slightly_smiling_face:

gkalpak replied:
Yeah, I was hoping that `@typescript-eslint` or `@angular-eslint` could include the necessary dependencies and we wouldn't have to bring in (and have to manage) a ton of them 🤷‍♂️

dario-piotrowicz replied:
I removed the schematics

I am honestly not sure about the others, I will look into whether they are necessary or not

(@alan-agius4 did you comment that some are not?)

dario-piotrowicz replied:
Here's a description of the dependencies: https://github.com/angular-eslint/angular-eslint#packages-included-in-this-project

I have read them the descriptions and had a (very) quick look at the sources and I do think that:
 - `builder` is necessary (to actually execute eslint)
 - `eslint-plugin` is necessary for the `@angular-eslint` rules (we have them)
 - `template-parser` is necessary for the parsing of the templates (which we have rules for)
 - `eslint-plugin-template` is necessary for the `@angular-eslint/template` rules (we have them)
 - `schematics`, I actually removed them but maybe they could be useful for maintaining eslint in the project? :thinking:

dario-piotrowicz commented on aio/package.json:
```diff
@@ -114,21 +119,25 @@
     "esbuild": "^0.12.0",
-    "eslint": "^7.23.0",
+    "eslint": "^7.26.0",
+    "eslint-plugin-import": "latest",
```
my bad I was warned about this by the schematics but forgot to fix it! sorry! :sob:

dario-piotrowicz commented on aio/package.json:
```diff
@@ -178,4 +186,4 @@
     "yargs": "^17.0.0"
   }
-}
+}
```
it wasn't me, it was the schematics :see_no_evil:

dario-piotrowicz commented on aio/src/404-body.html:
```diff
@@ -27,7 +27,7 @@
             <div class="content">
               <div class="nf-container l-flex-wrap flex-center">
-                <img src="assets/images/support/angular-404.svg" width="300" height="300" />
+                <img src="assets/images/support/angular-404.svg" width="300" height="300" alt="not-found"/>
```
ah yeah good point I will amend the md too :slightly_smiling_face: 

about the dash.... yeah sorry, force of habit I guess :sweat_smile:

dario-piotrowicz commented on aio/src/app/custom-elements/resource/resource-list.component.ts:
```diff
@@ -4,7 +4,7 @@ import { Category } from './resource.model';
 import { LocationService } from 'app/shared/location.service';
 
-/* tslint:disable:template-accessibility-elements-content */
+/* eslint-disable @angular-eslint/template/accessibility-elements-content */
```
sorry! I did remove it and forgot to push the change! on it! :stuck_out_tongue_winking_eye:

dario-piotrowicz commented on aio/src/app/app.component.spec.ts:
```diff
@@ -1311,7 +1311,7 @@ class TestHttpClient {
   ];
 
-  // tslint:disable:quotemark
+  /* eslint-disable quotes, @typescript-eslint/quotes */
```
yeah, this was done automatically and I didn't question it, it indeed doesn't make much sense... sorry :sweat:  I'll remove them anyways :slightly_smiling_face:

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,97 @@
+{
+  "root": true,
+  "ignorePatterns": [
+    "projects/**/*"
```
the schematics mention something about `projects/`: https://github.com/angular-eslint/angular-eslint#step-2---run-the-convert-tslint-to-eslint-schematic-on-a-project

I guess this is done by default so that if we did have multiple projects eslint wouldn't be applied to all of them

Sorry I didn't notice this, I'll get rid of it :slightly_smiling_face:

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,97 @@
+      }
+    }
+  ]
+}
```
makes sense! I'll get right on it! :slightly_smiling_face:

dario-piotrowicz commented on aio/src/app/custom-elements/code/code-example.component.ts:
```diff
@@ -67,11 +67,13 @@ export class CodeExampleComponent implements AfterViewInit {
   get hidecopy(): boolean { return this._hidecopy; }
   private _hidecopy: boolean;
 
+  /* eslint-disable quotes, @angular-eslint/no-input-rename */
```
sorry it was automated the `quotes` one is indeed unnecessary

Also I just saw that all automated `eslint-disable`s lack the `next-line` bit, I will amend that :disappointed:

dario-piotrowicz replied:
actually I think only the two `eslint-disable`s in this file lacked the `next-line` :sweat_smile:

dario-piotrowicz commented on aio/src/app/custom-elements/code/code-example.component.ts:
```diff
@@ -67,11 +67,13 @@ export class CodeExampleComponent implements AfterViewInit {
     this.hidecopy = hidecopy;
   }
 
+  /* eslint-disable quotes, @angular-eslint/no-input-rename */
```
same as above I'll fix it :slightly_smiling_face:

dario-piotrowicz commented on aio/src/app/custom-elements/code/code.component.spec.ts:
```diff
@@ -290,7 +290,7 @@ describe('CodeComponent', () => {
 
 //// Test helpers ////
-// tslint:disable:member-ordering
+/* eslint-disable @typescript-eslint/member-ordering */
```
sorry what does? isn't this just a one-to-one conversion of the already existing rule? :sweat:

gkalpak replied:
Yeah, I mean the already existing rule was redundant (and by extention the updated rule is redundant as well :grin:)

dario-piotrowicz replied:
ah ok :slightly_smiling_face: , removed :+1:

dario-piotrowicz commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,97 @@
+      "parserOptions": {
+        "project": [
+          "tsconfig.json",
+          "e2e/tsconfig.json"
```
yeah, this was automated, I'm on it :slightly_smiling_face:

alan-agius4 commented on aio/package.json:
```diff
@@ -114,21 +119,25 @@
     "esbuild": "^0.12.0",
-    "eslint": "^7.23.0",
+    "eslint": "^7.26.0",
+    "eslint-plugin-import": "^2.23.4",
```
This seems to be unused.

alan-agius4 commented on aio/package.json:
```diff
@@ -114,21 +119,25 @@
+    "eslint": "^7.26.0",
+    "eslint-plugin-import": "^2.23.4",
     "eslint-plugin-jasmine": "^4.1.2",
+    "eslint-plugin-jsdoc": "^35.4.3",
```
Also this one.

alan-agius4 commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,94 @@
+{
```
I am seeing a lot of missing rules from here. Was the tslint.json migrated using the migration tool
[Uploading Alan_Agius_-_Alan_Agius.pdf…]()
?

Example, `ban`, `no-console` rules are missing.

dario-piotrowicz replied:
There isn't an eslint `ban` rule, I used the `no-restricted-syntax` to accomplish the same thing here

gkalpak commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,94 @@
+        ],
+        "dot-notation": "error",
+        "indent": "off",
+        "@typescript-eslint/member-delimiter-style": ["error", {
```
:grin:

```suggestion
        "@typescript-eslint/member-delimiter-style": ["error", {
          "singleline": {
            "delimiter": "comma",
            "requireLast": false
          }
        }],
```

dario-piotrowicz replied:
:man_facepalming: sorry

gkalpak commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,94 @@
+        "prefer-arrow/prefer-arrow-functions": "off",
+        "quotes": "off",
+        "@typescript-eslint/quotes": ["error", "single", {"avoidEscape": true}],
+        "semi": "error",
+        "singleline": {
+          "delimiter": "comma",
+          "requireLast": false
+        }
+        }]
```
:grin:

```suggestion
        "semi": "error"
```

gkalpak reviewed (changes requested):
Thx for making the changes, @dario-piotrowicz :+1:
Getting closer 🏃‍♂️

> Well, your comments don't seem to be "just a few" 😅

Given the size of the PR, they were really "just a few" 😛

> I really hope I am not burdening you with this 😓
> thank you so very much for taking the time to review the PR 🥰

It would have taken me much more time to do the actual work myself, so thank _you_ for taking it on 💪

> By the way, could you comment on the typescript version warning? don't you see it as a problem here? 🙂

Sorry, I meant to comment on that, but then forgot.
Yeah, it's not ideal, but I don't think it is a problem here (as long as linting passes 😅).

##
<sub>**[NOTE-TO-SELF]** Unresolved comments from previous review:</sub>
<sub>https://github.com/angular/angular/pull/42820#discussion_r668897702, https://github.com/angular/angular/pull/42820#discussion_r668900621, https://github.com/angular/angular/pull/42820#discussion_r668906308, https://github.com/angular/angular/pull/42820#discussion_r668907227, https://github.com/angular/angular/pull/42820#discussion_r668908351, https://github.com/angular/angular/pull/42820#discussion_r668998288</sub>

dario-piotrowicz commented:
Just checked the `eslint-disable`s without `next-line` there were just two :slightly_smiling_face:

dario-piotrowicz commented:
@gkalpak I think that it's looking good :smiley: 

I was about to put this as ready for review but I noticed that the `test_aio` is failing and it is because of the `example-lint` and `tools-lint` scripts which don't work anymore as they use tslint which has been removed as part of the migration

I did not notice those before (and I am not sure what they are about), I will have a look tomorrow, hopefully once these are taken care of we will be in a good place with this :slightly_smiling_face:

gkalpak commented:
> I was about to put this as ready for review but I noticed that the `test_aio` is failing and it is because of the `example-lint` and `tools-lint` scripts which don't work anymore as they use tslint which has been removed as part of the migration

Ah, good point. For now, I think it is better to just focus on migrating the AIO app to eslint and keep tslint for the other scripts that need it (to keep the PR scope manageable). In a follow-up PR we can look into switching to eslint for more AIO scripts.

dario-piotrowicz commented:
All right @gkalpak !!! :rocket: 

I reinstated tslint and codelyzer and all seems to pass (the ci was a bit flaky and I needed to re-run it a couple of times :sweat_smile:)

it's looking quite good I think/hope :smile:

Today after work I'll put there a good comprehensive commit message and I will move it to ready for review :slightly_smiling_face:

dario-piotrowicz committed:
refactor(docs-infra): migrate aio from tslint to eslint

migrate aio to eslint as tslint has been deprecated, the migration is restricted to the aio app and
its e2e tests and does not include the other tools, for such reason both tslint and codelyzer have not
been removed (to be done in a next PR)

some minor tweaks needed to be applied to the code so that it would adhere to the new ESLinting behaviour

most TSLint rules have been substituted with their ESLint equivalent, with some exceptions:
  * [whitespace] does not have an ESLint equivalent (suggested to be handled by prettier)
  * [import-spacing] does not have an ESLint equivalent (suggested to be handled by prettier)
  * [ban] replaced with [no-restricted-syntax] as there is no (official/included) ESLint equivalent

some rules have minor different behaviours compared to their TSLint counterparts:
  * @typescript-eslint/naming-convention:
    - typescript-eslint does not enforce uppercase for const only.
  * @typescript-eslint/no-unused-expressions:
    - The TSLint optional config "allow-new" is the default ESLint behavior and will no longer be ignored.
  * arrow-body-style:
    - ESLint will throw an error if the function body is multiline yet has a one-line return on it.
  * eqeqeq:
    - Option "smart" allows for comparing two literal values, evaluating the value of typeof and null comparisons.
  * no-console:
    - Custom console methods, if they exist, will no longer be allowed.
  * no-invalid-this:
    - Functions in methods will no longer be ignored.
  * no-underscore-dangle:
    - Leading and trailing underscores (_) on identifiers will now be ignored.
  * prefer-arrow/prefer-arrow-functions:
    - ESLint does not support allowing standalone function declarations.
    - ESLint does not support allowing named functions defined with the function keyword.
  * space-before-function-paren:
    - Option "constructor" is not supported by ESLint.
    - Option "method" is not supported by ESLint.

additional notes:
  * the current typescript version used by the aio app is 4.3.5, which is not supported by typescript-eslint (the supported
    versions are >=3.3.1 and <4.3.0). this causes a warning message to appear during linting, this issue should
    likely/hopefully disappear in the future as typescript-eslint catches up
  * The new "no-console" rule is not completely equivalent to what we had prior the migration, this is because TSLint's "no-console"
    rule let you specify the methods you did not want to allow, whilst ESLint's "no-console" lets you specify the methods that you do
    want to allow, so and in order not to have a very long list of methods in the ESLint rule it's been decided for the time being
    to simply only allow the "log", "warn" and "error" methods
  * 4 dependencies have been added as they have been considered necessary (see: https://github.com/angular/angular/pull/42820#discussion_r669978232)

extra:
  * the migration has been performed by following: https://github.com/angular-eslint/angular-eslint#migrating-an-angular-cli-project-from-codelyzer-and-tslin
  * more on typescript-eslint at: https://github.com/typescript-eslint/typescript-eslint

dario-piotrowicz commented:
So commit message updated and PR ready for review, hopefully we're in a good spot now and we're getting close! :slightly_smiling_face: 

@rafaelss95 and @alan-agius4  thanks a lot for your nice reviewing I really appreciate it!!! :slightly_smiling_face: 
( @rafaelss95 not sure why I am not able to re-request your review :confused: )
(please feel free to check again the PR with the latest changes if you'd like)

And of course @gkalpak can't really thank you enough for all the support you always give me!!!! :smiling_face_with_three_hearts:

gkalpak commented on aio/tests/e2e/src/app.e2e-spec.ts:
```diff
@@ -230,15 +230,16 @@ describe('site App', () => {
       await page.navigateTo('guide/router');
       expect(await page.ghLinks.count()).toEqual(1);
-      /* tslint:disable:max-line-length */
+
```
Nit: 

```suggestion
```

gkalpak commented on aio/tests/deployment/e2e/site.po.ts:
```diff
@@ -34,6 +34,7 @@ export class SitePage {
     const results = element.all(by.css('.search-results li'));
     await browser.wait(ExpectedConditions.presenceOf(results.first()), 8000);
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
     return await results.map<string>(link => link!.getText());
```
I think you can also do `(link as ElementFinder).getText()` or even `link?.getText()` and avoid disabling the lint rule.

gkalpak reviewed (approved):
Looks awesome 🏆
Thank you so much for taking care of this, @dario-piotrowicz 😍

And the detailed commit message is super helpful too :100:

##
One last question regarding:
> for some reason the "no-console" rule would not work in eslint and has been substituted with an equivalent "no-restricted-syntax" one

What exactly didn't work with the `no-console` rule? (I've never had problems with it in other projects 😕)

##
Reviewed-for: global-docs-approvers

dario-piotrowicz commented:
> Looks awesome 
> Thank you so much for taking care of this, @dario-piotrowicz 
> 
> And the detailed commit message is super helpful too 
> 
> One last question regarding:
> 
> > for some reason the "no-console" rule would not work in eslint and has been substituted with an equivalent "no-restricted-syntax" one
> 
> What exactly didn't work with the `no-console` rule? (I've never had problems with it in other projects )
> 
> Reviewed-for: global-docs-approvers

thanks!!!!! :heart_eyes: 

about the `"no-console"`, sorry that's just another silly mistake of mine :see_no_evil: :sweat_smile: 

I just checked and it does work, the issue now is that the eslint rule has the `allows` option in which you specify which methods you want to allow, whilst the tslint rule made you specify which you didn't want to allow.

so with `no-restricted-syntax` I am keeping the same exact behaviour here, if I had to do the same with `no-console` I think I'd have to specify quite some methods :sad (https://developer.mozilla.org/en-US/docs/Web/API/Console#methods)

Alternatively I can just put `no-console` and only allow `log`, `warn` and `error`, that should also be pretty similar I think

what do you think the best solution here would be? :slightly_smiling_face:

gkalpak reviewed (commented):
Cool! I think enabling the `no-console` rule and only allowing `log`, `warn` and `error` is fine. (We could also consider `info` and `debug`, but they are not used anywhere, we can leave them out for now.)

dario-piotrowicz commented:
@gkalpak should be all good now :smiley:

gkalpak commented on aio/.eslintrc.json:
```diff
@@ -0,0 +1,91 @@
+        }],
+        "@typescript-eslint/member-ordering": "off",
+        "@typescript-eslint/naming-convention": "off",
+        "no-console": [ "error", {"allow": ["log", "warn", "error"]} ],
```
Super-nit: Let's be consistent with whitespace around array brackets:

```suggestion
        "no-console": ["error", {"allow": ["log", "warn", "error"]}],
```

##
Meta-question: Who lints your lint-config 🤔 😛

dario-piotrowicz replied:
sorry :bow: 

About the linting on the lint-config, who needs that when we've got such an awesome reviewer!? :heart_eyes: :stuck_out_tongue_winking_eye: 

`gkalpakLint` for the win! :rocket:

gkalpak reviewed (approved):
:100: 🚀 :shipit:

alxhub commented:
I switched this to `target: minor` because it doesn't merge cleanly into the current patch branch (`12.1.x`). Given the nature of this change, I think this is probably fine.
